recipes-kernel/kernel-tests/kernel-tests-files/test_i915_firmware.sh

```diff
@@ -41,43 +44,50 @@ function check_files () {
 	done
 }
 
-# Use dmidecode to check the system-manufacturer is 'National Instruments'.
-# If it is not, skip this test.
-function check_manufacturer () {
-	MAN=$(dmidecode --string system-manufacturer) \
-	|| echo "ERROR: dmicode call failed." 
+function check_video_device () {
+	has_video_device=0
+	has_intel_video_device=0
+	has_i915_driver_loaded=0
 
-	if [ ! "${MAN}" == "National Instruments" ]; then
-		echo "REASON: Detected system manufacturer as '${MAN}'."
-		echo "REASON: We do not know if i915 should be tested."
-		ptest_skip
-	fi
-}
+	# Enumerate PCI devices, looking for video controllers (that may
+	# or may not have drivers bound to them)
+	while read dev; do
+		pci_class=$(cat $SYSFS_PCI_DEVS/$dev/class)
+		# device class 0x03 is "display controller"
+		pci_class_id=$(( ($pci_class & 0xFF0000) >> 16 ))
+		if [ $pci_class_id -eq 3 ]; then
+			has_video_device=1
 
-# Passes iff the i915 module is loaded AND has >0 users. Fails otherwise.
-function check_module () {
-	MOD=$(lsmod | awk '/^i915/ {print $3}')
-	if [ -z "$MOD" ]; then
-		echo "i915 module not loaded."
-		ptest_fail
-		return
-	fi
+			# get the vendor id
+			vendor_id=$(cat $SYSFS_PCI_DEVS/$dev/vendor)
+			if [ "${vendor_id}" = "0x8086" ]; then
+				has_intel_video_device=1
+			fi
+
+			driver_name=$(basename $(realpath $SYSFS_PCI_DEVS/$dev/driver))
+			if [ "${driver_name}" = "i915" ]; then
+				has_i915_driver_loaded=1
```

Comment:
Do you need to continue looking at the remaining devices once you find the i915 driver?

Reply:
Nope, we do not. My initial stab at this loop didn't incorporate the driver check because I didn't expect to remove `check_module`, but yeah I think now once we satisfy the precondition of "we have at least one GPU device with i915 loaded" we can stop looking.